HealthGPT/HealthGPT/HealthDataFetcher.swift

```diff
@@ -64,120 +67,151 @@ class HealthDataFetcher {
             intervalComponents: DateComponents(day: 1)
         )
 
-        var dailyData: [Double] = []
-
-        query.initialResultsHandler = { _, results, _ in
-            if let statsCollection = results {
-                statsCollection.enumerateStatistics(
-                    from: Date().twoWeeksAgoStartOfDay(),
-                    to: Date.startOfDay()
-                ) { statistics, _ in
-                    if let quantity = statistics.sumQuantity() {
-                        dailyData.append(quantity.doubleValue(for: unit))
-                    } else {
-                        dailyData.append(0)
+        return try await withCheckedThrowingContinuation { continuation in
+            query.initialResultsHandler = { _, results, error in
+                if let error = error {
+                    continuation.resume(throwing: error)
+                } else if let statsCollection = results {
+                    var dailyData: [Double] = []
+
+                    statsCollection.enumerateStatistics(
+                        from: Date().twoWeeksAgoStartOfDay(),
+                        to: Date.startOfDay()
+                    ) { statistics, _ in
+                        if let quantity = statistics.sumQuantity() {
+                            dailyData.append(quantity.doubleValue(for: unit))
+                        } else {
+                            dailyData.append(0)
+                        }
                     }
-                }
 
-                completion(dailyData)
-            } else {
-                completion([])
+                    continuation.resume(returning: dailyData)
+                } else {
+                    continuation.resume(throwing: HealthDataFetcherError.resultsNotFound)
+                }
             }
-        }
 
-        healthStore.execute(query)
+            healthStore.execute(query)
+        }
```

Comment:
While this is a valid and functionally correct way to implement it. Apple actually provides even more convenient async overloads for the functions that you can use here. You can find a good example using `HKStatisticsCollectionQueryDescriptor` here: https://developer.apple.com/documentation/healthkit/hkstatisticscollectionquerydescriptor

Reply:
Thank you @PSchmiedmayer, I've implemented this!